### PR TITLE
Merge gio document loader and saver into document loader and saver

### PR DESCRIPTION
**Pull request by mbkma**
Backport from: https://gitlab.gnome.org/GNOME/gedit/-/commit/4bd74a1f47a3fa41385ffae3bb78aeb5afabb564
See: https://bugzilla.gnome.org/show_bug.cgi?id=617215

I think I need this backport for a future PR I want to make. Apparently having a separate gio-document-loader/saver is not needed anymore.

### CODE REVIEW

**Comment by raveit65** [comment]
It seems that everything works like before in a quick test.

File: pluma/pluma-document-loader.c [hunk 1]
```diff
@@ -218,8 +302,502 @@ pluma_document_loader_class_init (PlumaDocumentLoaderClass *klass)
...
+
+	loader->priv->used = FALSE;
+	loader->priv->auto_detected_newline_type = PLUMA_DOCUMENT_NEWLINE_TYPE_DEFAULT;
+	loader->priv->converter = NULL;
+	loader->priv->error = NULL;
+}
+
+PlumaDocumentLoader *
+pluma_document_loader_new (PlumaDocument       *doc,
+			   const gchar         *uri,
```
Comment (raveit65) [inline pluma/pluma-document-loader.c:315]: Indents

File: pluma/pluma-document-loader.c [hunk 2]
```diff
@@ -218,8 +302,502 @@ pluma_document_loader_class_init (PlumaDocumentLoaderClass *klass)
...
+static void
+async_failed (AsyncData *async, GError *error)
+{
+	g_propagate_error (&async->loader->priv->error, error);
+	remote_load_completed_or_failed (async->loader, async);
+}
+
+static void
+close_input_stream_ready_cb (GInputStream *stream,
+			     GAsyncResult  *res,
```
Comment (raveit65) [inline pluma/pluma-document-loader.c:409]: Indents

File: pluma/pluma-document-loader.c [hunk 3]
```diff
@@ -218,8 +302,502 @@ pluma_document_loader_class_init (PlumaDocumentLoaderClass *klass)
...
+	pluma_document_loader_loading (loader,
+				       FALSE,
+				       NULL);
+
+	read_file_chunk (async);
+}
+
+static void
+async_read_cb (GInputStream *stream,
+	       GAsyncResult *res,
```
Comment (raveit65) [inline pluma/pluma-document-loader.c:495]: Indents

File: pluma/pluma-document-loader.c [hunk 4]
```diff
@@ -218,8 +302,502 @@ pluma_document_loader_class_init (PlumaDocumentLoaderClass *klass)
...
+	/* Output stream */
+	loader->priv->output = pluma_document_output_stream_new (loader->priv->document);
+
+	/* start reading */
+	read_file_chunk (async);
+}
+
+static void
+query_info_cb (GFile        *source,
+	       GAsyncResult *res,
```
Comment (raveit65) [inline pluma/pluma-document-loader.c:653]: Indents

File: pluma/pluma-document-loader.c [hunk 5]
```diff
@@ -218,8 +302,502 @@ pluma_document_loader_class_init (PlumaDocumentLoaderClass *klass)
...
+	}
+
+	async->loader->priv->info = info;
+
+	finish_query_info (async);
+}
+
+static void
+mount_ready_callback (GFile        *file,
+		      GAsyncResult *res,
```
Comment (raveit65) [inline pluma/pluma-document-loader.c:687]: Indents

File: pluma/pluma-document-loader.c [hunk 6]
```diff
@@ -218,8 +302,502 @@ pluma_document_loader_class_init (PlumaDocumentLoaderClass *klass)
...
+				       async->cancellable,
+				       (GAsyncReadyCallback) mount_ready_callback,
+				       async);
+
+	g_object_unref (mount_operation);
+}
+
+static void
+async_read_ready_callback (GObject      *source,
+			   GAsyncResult *res,
```
Comment (raveit65) [inline pluma/pluma-document-loader.c:739]: Indents

File: pluma/pluma-document-loader.c [hunk 7]
```diff
@@ -218,8 +302,502 @@ pluma_document_loader_class_init (PlumaDocumentLoaderClass *klass)
...
+
+	/* get the file info: note we cannot use
+	 * g_file_input_stream_query_info_async since it is not able to get the
+	 * content type etc, beside it is not supported by gvfs.
+	 * Using the file instead of the stream is slightly racy, but for
+	 * loading this is not too bad...
+	 */
+	g_file_query_info_async (loader->priv->gfile,
+				 REMOTE_QUERY_ATTRIBUTES,
+                                 G_FILE_QUERY_INFO_NONE,
```
Comment (raveit65) [inline pluma/pluma-document-loader.c:786]: Indents

File: pluma/pluma-document-saver.c [hunk 8]
```diff
@@ -67,6 +66,52 @@ enum {
...
+	gchar 		       buffer[WRITE_CHUNK_SIZE];
+	GCancellable 	      *cancellable;
+	gboolean	       tried_mount;
+	gssize		       written;
+	gssize		       read;
+	GError                *error;
+} AsyncData;
+
+#define REMOTE_QUERY_ATTRIBUTES G_FILE_ATTRIBUTE_STANDARD_CONTENT_TYPE "," \
+				G_FILE_ATTRIBUTE_TIME_MODIFIED "," \
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:81]: Indents

File: pluma/pluma-document-saver.c [hunk 9]
```diff
@@ -243,52 +348,582 @@ pluma_document_saver_new (PlumaDocument           *doc,
...
+						   "encoding", encoding,
+						   "newline_type", newline_type,
+						   "flags", flags,
+						   NULL));
+
+}
+
+static void
+remote_save_completed_or_failed (PlumaDocumentSaver *gvsaver,
+				 AsyncData 	       *async)
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:364]: Indents

File: pluma/pluma-document-saver.c [hunk 10]
```diff
@@ -243,52 +348,582 @@ pluma_document_saver_new (PlumaDocument           *doc,
...
+	cancel_output_stream (async);
+}
+
+/*
+ * END NOTE
+ */
+
+static void
+remote_get_info_cb (GFile        *source,
+		    GAsyncResult *res,
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:462]: Indents

File: pluma/pluma-document-saver.c [hunk 11]
```diff
@@ -243,52 +348,582 @@ pluma_document_saver_new (PlumaDocument           *doc,
...
+	remote_save_completed_or_failed (saver, async);
 }
 
-void
-pluma_document_saver_saving (PlumaDocumentSaver *saver,
-			     gboolean            completed,
-			     GError             *error)
+static void
+close_async_ready_get_info_cb (GOutputStream *stream,
+			       GAsyncResult  *res,
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:501]: Indents

File: pluma/pluma-document-saver.c [hunk 12]
```diff
@@ -243,52 +348,582 @@ pluma_document_saver_new (PlumaDocument           *doc,
...
+}
 
-	if (completed)
+/* prototype, because they call each other... isn't C lovely */
+static void read_file_chunk (AsyncData *async);
+static void write_file_chunk (AsyncData *async);
+
+static void
+async_write_cb (GOutputStream *stream,
+		GAsyncResult  *res,
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:571]: Indents

File: pluma/pluma-document-saver.c [hunk 13]
```diff
@@ -243,52 +348,582 @@ pluma_document_saver_new (PlumaDocument           *doc,
...
+	/* Get how many chars have been read */
+	dstream = PLUMA_DOCUMENT_INPUT_STREAM (saver->priv->input);
+	saver->priv->bytes_written = pluma_document_input_stream_tell (dstream);
+
+	write_file_chunk (async);
+}
+
+static void
+async_replace_ready_callback (GFile        *source,
+			      GAsyncResult *res,
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:678]: Indents

File: pluma/pluma-document-saver.c [hunk 14]
```diff
@@ -243,52 +348,582 @@ pluma_document_saver_new (PlumaDocument           *doc,
...
+		g_object_unref (file_stream);
+		g_object_unref (converter);
+	}
+	else
+	{
+		saver->priv->stream = G_OUTPUT_STREAM (file_stream);
+	}
+
+	saver->priv->input = pluma_document_input_stream_new (GTK_TEXT_BUFFER (saver->priv->document),
+								saver->priv->newline_type);
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:727]: Indents

File: pluma/pluma-document-saver.c [hunk 15]
```diff
@@ -243,52 +348,582 @@ pluma_document_saver_new (PlumaDocument           *doc,
...
+			      G_FILE_CREATE_NONE,
+			      G_PRIORITY_HIGH,
+			      async->cancellable,
+			      (GAsyncReadyCallback) async_replace_ready_callback,
+			      async);
+}
+
+static void
+mount_ready_callback (GFile        *file,
+		      GAsyncResult *res,
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:766]: Indents

File: pluma/pluma-document-saver.c [hunk 16]
```diff
@@ -243,52 +348,582 @@ pluma_document_saver_new (PlumaDocument           *doc,
...
+				       async->cancellable,
+				       (GAsyncReadyCallback) mount_ready_callback,
+				       async);
+
+	g_object_unref (mount_operation);
+}
+
+static void
+check_modification_callback (GFile        *source,
+			     GAsyncResult *res,
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:818]: Indents

File: pluma/pluma-document-saver.c [hunk 17]
```diff
@@ -243,52 +348,582 @@ pluma_document_saver_new (PlumaDocument           *doc,
...
+			pluma_debug_message (DEBUG_SAVER, "Error getting modification: %s", error->message);
+
+			async_failed (async, error);
+			return;
+		}
+	}
+
+	/* check if the mtime is > what we know about it (if we have it) */
+	if (info != NULL && g_file_info_has_attribute (info,
+				       G_FILE_ATTRIBUTE_TIME_MODIFIED))
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:857]: Indents

File: pluma/pluma-document-saver.c [hunk 18]
```diff
@@ -298,39 +933,75 @@ pluma_document_saver_save (PlumaDocumentSaver     *saver,
...
 	// async is extra work here, sync is special casing in the caller
 
 	/* never keep backup of autosaves */
-	if ((saver->flags & PLUMA_DOCUMENT_SAVE_PRESERVE_BACKUP) != 0)
-		saver->keep_backup = FALSE;
+	if ((saver->priv->flags & PLUMA_DOCUMENT_SAVE_PRESERVE_BACKUP) != 0)
+		saver->priv->keep_backup = FALSE;
 	else
-		saver->keep_backup = pluma_prefs_manager_get_create_backup_copy ();
+		saver->priv->keep_backup = pluma_prefs_manager_get_create_backup_copy ();
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:950]: Indent

File: pluma/pluma-document-saver.c [hunk 19]
```diff
@@ -298,39 +933,75 @@ pluma_document_saver_save (PlumaDocumentSaver     *saver,
...
+	g_timeout_add_full (G_PRIORITY_HIGH,
+			    0,
+			    (GSourceFunc) save_remote_file_real,
+			    saver,
+			    NULL);
+}
+
+void
+pluma_document_saver_saving (PlumaDocumentSaver *saver,
+			     gboolean            completed,
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:967]: Indents

**Comment by mbkma** [comment]
@raveit65 It only does look weird on Github, not in Pluma (which may be caused bay tabs + spaces, but this was the original coding style)
Screenshot at 2020-07-15 13-28-33

**Comment by raveit65** [comment]
Da hast Du leider Pech gehabt das du alten schlecht formatierten Code angefasst hast :)
Nun ist es aber neu geschriebener Code.
Es wäre schön wenn es auch auf github schön aussieht.
Danke

**Comment by mbkma** [comment]
@raveit65 this was the most difficult thing I've ever done ;)

File: pluma/pluma-document-saver.c [hunk 20]
```diff
@@ -49,334 +49,334 @@
...
+                                     PROP_URI,
+                                     g_param_spec_string ("uri",
+                                     "URI",
+                                     "The URI this PlumaDocumentSaver saves the document to",
+                                     "",
+                                     G_PARAM_READWRITE |
+                                     G_PARAM_CONSTRUCT_ONLY |
+                                     G_PARAM_STATIC_STRINGS));
+
+    g_object_class_install_property (object_class,
```
Comment (raveit65) [inline pluma/pluma-document-saver.c:290]: Can you please re-check line 290 to 310 (2 blocks) ?
Rest looks Ok.
Thanks

Comment (mbkma) [reply pluma/pluma-document-saver.c:290]: should be fine now.

File: pluma/pluma-document-loader.c [hunk 21]
```diff
@@ -72,864 +72,862 @@ static guint signals[LAST_SIGNAL] = { 0 };
...
 }
 
 static void
 remote_load_completed_or_failed (PlumaDocumentLoader *loader, AsyncData *async)
 {
-	pluma_document_loader_loading (loader,
-				       TRUE,
-				       loader->priv->error);
+    pluma_document_loader_loading (loader,
+                       TRUE,
```
Comment (raveit65) [inline pluma/pluma-document-loader.c:393]: Oh no, indent....

**Comment by raveit65** [comment]
One indent issue is missing :) 
Any way, i will start to test the resulting patch again.
Btw. when adding `.diff` or `.patch` to the end of PR link in browser you see the raw formatted version.

File: pluma/pluma-document-loader.c (hunk 21 above)
Comment (mbkma) [reply pluma/pluma-document-loader.c]: done :)

**Comment by raveit65** [comment]
Am i right that behaviour of pluma isn't change with this PR?
In result all is fine when it works like before?

**Comment by mbkma** [comment]
Yes nothing should have changed. I just merged the abstract class into the concrete class, without actual code changes. I don't even know why the abstraction was needed, but it seems it [is not needed anymore](https://bugzilla.gnome.org/show_bug.cgi?id=617215).

**Comment by mbkma** [comment]
Today I was testing this PR again with `diff` all files compared to master branch. I overlooked this one mate-commit. Now everything should be identical. I can squash commit before merge.

File: pluma/pluma-document-saver.c [hunk 22]
```diff
@@ -31,329 +31,997 @@
...
+    g_return_if_fail (PLUMA_IS_DOCUMENT_SAVER (saver));
+    g_return_if_fail (saver->priv->uri != NULL && strlen (saver->priv->uri) > 0);
+
+    g_return_if_fail (saver->priv->used == FALSE);
+    saver->priv->used = TRUE;
+
+    // CHECK:
+    // - sanity check a max len for the uri?
+    // report async (in an idle handler) or sync (bool ret)
+    // async is extra work here, sync is special casing in the caller
```
Comment (rbuj) [inline pluma/pluma-document-saver.c:941]: Do no use cpp inline comments

**Pull request merged**